### DISCUSSION TITLE
Status Report Request

### DISCUSSION THREAD

**Discussion by gotjoshua** [zxch3n/PomodoroLogger#236] [General]
Greetings,
I found and [forked](https://github.com/onezoomin/PomodoroLogger/issues) your project today, and I am quite impressed with the code organization and core functionality! Thanks for sharing this with the FOSS community at large!

I made a [list of the first issues that I intend to address in my fork](https://github.com/onezoomin/PomodoroLogger/issues).

I am writing here to invite feedback and to request an official status update from @zxch3n  as the maintainer...
Do you have any intentions of picking up the project any time soon?

**Comment by zxch3n** [comment]
Cool! I'm happy to see that you want to refactor the persistence layer.

I can't spend too much time on this project lately. But I'd love to review your PRs. 

Did you consider using something like pouchDB as the new db? It works on both browser and node.

**Reply by gotjoshua** [reply]
> something like pouchDB

i prefer dexie (uses indexDB) as the data and api are more clean and simple. 

there is a very well maintained fork of neDB, so if we were to keep the neDB layer at all, i'd at least want to migrate to that.

**Reply by zxch3n** [reply]
Which fork? 

I'm worried about using indexedDB. I'm not sure whether reinstalling/deleting the app then installing/upgrading electron would cause potential data lost.

**Reply by gotjoshua** [reply]
> Which fork?

https://github.com/seald/nedb/network

So... Indexdb in electron is new for me, i need to look into the details. Thanks for the heads up!

> works on both browser and node.

Is there an actual node.js requirement at the moment?

**Reply by zxch3n** [reply]
I found [an article about potential data loss](https://dexie.org/docs/StorageManager#:~:text=Even%20though%20IndexedDB%20is%20a,disk%20space%20on%20a%20device.).  And a discussion about [browsers may delete your IndexedDB database under certain conditions](https://news.ycombinator.com/item?id=28158407). But it seems like the problem is mainly on Safari's implementation. Maybe we are good on Chromium. 

Working on node.js can make tests easier, and it feels safer. But I'm happy to use a browser-only solution if the data safety is guaranteed.

**Reply by gotjoshua** [reply]
> I'm happy to use a browser-only solution if the data safety is guaranteed.

cool, i am into offline first, but not necessarily browser-only...

at the moment i don't see a need for node.js, but i wouldn't like to break the tests by switching to dexie. apparently there are ways to shim indexdb for node (which might be good enough for the tests)

the dexie article you linked to doesn't mention electron, so i think we need to research more about "persistence" of indexdb in electron specifically.

also, my first thought was to add dexie and keep nedb (why change a working system) but my top priority is gitlab integration.
I even consider a data mode that considers gitlab issues to be the source of truth and uses dexie as offline-first and reactive cache

I will try to include flexible settings in each PR (if/when i get that far) so that each new feature proposal will be opt in via the prefs page.

**Reply by zxch3n** [reply]
I also prefer offline-first. I'd love it to be [local-first](https://www.inkandswitch.com/local-first/). I may integrate CRDT someday, but that's another topic.

I guess things that works in Chromium probably works on electron as well. 

 Feature gating is cool. But it might be hard when you want to change db. 

For gitlab integration, it will be awesome if you could abstract a reusable integration layer. So that we can reuse it on github integration/notion integration.

**Reply by gotjoshua** [reply]
> I'd love it to be [local-first](https://www.inkandswitch.com/local-first/).

great article (btw) !

i am digging into the concept these days.



> abstract a reusable integration layer

sounds good, but i must admit it will take some time till i manage to give this priority

**Reply by zxch3n** [reply]
Yes. Take your time 😁

**Comment by zxch3n** [comment]
I converted this issue to a dicussion

**Reply by gotjoshua** [reply]
Nice, yeah that's more fitting for sure.

Also just want to say that it's great to see that you are still around and into discussions even if you don't have so much bandwidth for coding on this project these days.

**Reply by gotjoshua** [reply]
some progress...

https://github.com/onezoomin/PomodoroLogger/issues/2#issuecomment-1091733915

**Reply by zxch3n** [reply]
Great job!
